File: LeetCode/0317.Shortest-Distance-From-All-Buildings/Shortest-Distance-From-All-Buildings.java
class Solution{
    class Point{
         int x;
         int y;
         public Point(int x,int y){
              this.x = x;
              this.y = y;
          }
    }
    public int shortestDistance(int[][] grid) {
           int m = grid.length;
           int n = grid[0].length;
           int [][] dist = new int [m][n];
           //dist[1][2] = 7
           int [][] num = new int [m][n];
           //num[1][2] = 3
           int totalBuilding = 0;
           for(int i = 0; i < m; i++){
                 for(int j = 0; j < n; j++){
                         if(grid[i][j] == 1){
                               totalBuilding++;
                               bfs(grid, i, j, dist, num);
                          }
                  }
            }
            int res = Integer.MAX_VALUE;
            for(int i = 0; i < m; i++){
                 for(int j = 0; j < n; j++){
                         if(num[i][j] == totalBuilding){
                                 res = Math.min(res, dist[i][j]);
                         }
                 }
             }
            if(res == Integer.MAX_VALUE) return -1;
            else return res;


    }
    public void bfs(int [][]grid, int row, int col, int [][]dist, int [][]num){

          Queue<Point> queue = new LinkedList<>();
          queue.offer(new Point(col, row));
          boolean [][] visited = new boolean[grid.length][grid[0].length];
          visited[row][col] = true;
          int [] dx = {1, -1, 0, 0};
          int [] dy = {0,  0, 1, -1};
          int distance = 0;
          while(!queue.isEmpty()){
                int size = queue.size();
                distance++;
                for(int i = 0; i < size; i++){
                      Point cur = queue.poll();
                      Point newP = new Point(0, 0);
                      for(int j = 0; j < 4; j++){
                            newP.x = cur.x + dx[j];
                            newP.y = cur.y + dy[j];
                            if(newP.x >= 0 && newP.x < grid[0].length && newP.y >= 0 && newP.y < grid.length && !visited[newP.y][newP.x] && grid[newP.y][newP.x] == 0){
                                    dist[newP.y][newP.x] += distance;
                                    visited[newP.y][newP.x] = true;
                                    num[newP.y][newP.x] ++;
                                    queue.offer(new Point(newP.x, newP.y));
                             }
                      }
                }
                
          }
     }
}

File: LeetCode/0317.Shortest-Distance-From-All-Buildings/Shortest-Distance-From-All-Buildings.py
class Solution(object):
    def shortestDistance(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        res = float('inf')
        m, n = len(grid), len(grid[0])
        empty = 0
        dist = [[0] * n for _ in range(m)]
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    res = float('inf')
                    queue = collections.deque()
                    queue.append((0, i, j))
                    while queue:
                        d, x, y = queue.popleft()
                        for nx, ny in [(x + 1, y), (x - 1, y), (x, y - 1), (x, y + 1)]:
                            if 0 <= nx < m and 0 <= ny < n and grid[nx][ny] == empty:
                                grid[nx][ny] -= 1
                                dist[nx][ny] += d + 1
                                res = min(res, dist[nx][ny])
                                queue.append((d + 1, nx, ny))
                    empty -= 1
        return -1 if res == float('inf') else res
